PBI-04 One Time Code

johaneschristian opened the pull request:
Create functionalities that allows companies to send one time code invitation links to assessors using Django mail. This commit also refactors tests and implementations that still contain code smells.

johaneschristian committed:
[REFACTOR] remove duplicated literal in tests.py

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-07-google_login

johaneschristian committed:
Merge branch 'staging' of https://github.com/one-day-intern/one-day-intern-backend into PBI-04-assessor_one_time_code

johaneschristian committed:
[RED] create stub and test for send_mass_html_mail

send_mass_html_mail is a function that will be used to send mass mail
with html as its content. This function will be build on top of django
mail module. This function is required as django's default mass mail
service does not allow for html content and only uses plain text. This
prevents us from including images and formatting the text payload.

johaneschristian committed:
[GREEN] implement send_mass_html_mail

Implement send_mass_html_mail using django EmailMultiAlternatives class.
Using the same connection also ensures a faster mail sending process.

johaneschristian committed:
[RED] create stub and test for get_company_or_raise_exception

The get_company_or_raise_exception will be used as a function to
properly authorize user in accessing resources related to company.
The function will return RestrictedAccessException when user is not a
company.

johaneschristian committed:
[GREEN] implement get_company_or_raise_exception

Implemented function used in authorizing request.user as company in
accessing company related resources.

johaneschristian committed:
[RED] create stubs and tests for one time code sending functionality

Create skeletons, stubs, and tests for functions that will be involved
in generating the one time code and sending it to the assessor's email.

johaneschristian committed:
[GREEN] implement service functionalities for sending OTC

johaneschristian committed:
[GREEN] fix test of one time code for assessor registration

A set of tests for assessor registration, involving one time code, has
been created earlier. However, two of the tests were problematic as it
attempts to produce a non registered code by replacing the last
character with the first character. Such approach may lead to
nondeterministic behavior as the last character may be equal to the
first one, hence resulting to the same code. This commit replaces it
with a new generated UUID that is guaranteed to be different.

johaneschristian committed:
[RED] create stub and test for serve_send_one_time_code_to_assessors

The serve_send_one_time_code_to_assessors will act as the controller
that handle the user's request. It will invoke the necessary service
functionalities to send the OTC to the list of assessors.

johaneschristian committed:
[GREEN] implement serve_send_one_time_code_to_assessors

Implement the function and modified exception_config to accomodate new
exceptions by returning the appropriate response status code depending
on the severity of the exception.

johaneschristian committed:
[REFACTOR] remove duplications in company and user tests

Remove duplications in company and user tests according to guidance
provided by sonarcloud. In this commit, repeated parts are separated by
defining constatns or extracting similar portions of the function.

johaneschristian committed:
[REFACTOR] remove unused imports from boiler plate

johaneschristian committed:
[REFACTOR] change tests that do not utilize self as static

rashadaziz commented on company/services/one_time_code.py:
```diff
@@ -0,0 +1,66 @@
+from users.models import Company, CompanyOneTimeLinkCode
+from users.services import utils as users_utils
+from . import utils, company as company_service
+from ..exceptions.exceptions import InvalidRequestException
```
Hey jo good job, but do you think it'll be better if you moved all the exceptions under one directory for easier maintainability and import path consistency? Let me know :D

johaneschristian replied:
Hello there, Rashad. That is an excellent suggestion. I'll start using it right away.

johaneschristian committed:
[REFACTOR] move all exceptions under exceptions.py

Before this commit, all of the exceptions are scattered across the
different applications. However, this implementation may result to
duplication as we often not know whether an exception already exists.
Hence, grouping the exception will give a better understanding on what's
available and what's not, and make it easier to maintain.
